[13.0][FIX] purchase_sale_stock_inter_company: Allow reception with serial tracking and partial quantity (Dropshipping)

carlos-lopez-tecnativa opened the pull request:
Issue: When handling intercompany dropshipping with serialized products, an error is raised during reception if the quantity received is less than the original purchase order's quantity. The error message states: "Mismatch between move lines with the corresponding."

**Steps to Reproduce:**

**Settings:**

In General Settings (both companies):
- Enable Sales from Purchase Orders.
- Disable Sales Order Auto Validation.
- In Purchase Settings: Enable Dropshipping.
- In Inventory Settings: Enable Multi-Step Routes.

![image](https://github.com/user-attachments/assets/b0bf7bad-10c5-4222-9821-69e45d697b9e)

**Setup:**

In Company 1:
Configure a product with vendor information linked to Company 2.
In Company 2:
Configure the same product with vendor information linked to any other vendor (not a company contact).
Scenario:

In Company 1:

- Create a sales order with a line using the Dropshipping route and the configured product.
- Confirm the sales order.
- Navigate to Purchase Orders, find the related record, and confirm it.

Switch to Company 2:

- Locate the generated sales order and update the sales line to use the Dropshipping route.
- Set the quantity to 2 and confirm the sales order.
- Go to Purchase Orders, find the related record, and confirm it.
- Navigate to Receipts for the purchase order, enter the lot/serial numbers, and attempt to validate the receipt.

Result: An error is raised, preventing the validation of the receipt due to a mismatch between the move lines and their corresponding records.

SO in C1
http://oca-multi-company-13-0-c48386e2793f.runboat.odoo-community.org/web#id=25&action=376&model=sale.order&view_type=form&cids=1&menu_id=224

PO in C1
http://oca-multi-company-13-0-c48386e2793f.runboat.odoo-community.org/web#id=12&action=355&model=purchase.order&view_type=form&cids=&menu_id=206

SO in C2
http://oca-multi-company-13-0-c48386e2793f.runboat.odoo-community.org/web#id=26&action=376&model=sale.order&view_type=form&cids=2&menu_id=224

PO in C2
http://oca-multi-company-13-0-c48386e2793f.runboat.odoo-community.org/web#id=13&action=355&model=purchase.order&view_type=form&cids=&menu_id=206

![image](https://github.com/user-attachments/assets/b4ddc754-54d7-4fc4-804b-0f32fbfd5b1a)


TT51858
@Tecnativa @pedrobaeza could you please review this

carlos-lopez-tecnativa committed:
[FIX] purchase_sale_stock_inter_company: Allow reception with serial tracking and partial quantity (Dropshipping)

Issue: When handling intercompany dropshipping with serialized products, an error is raised during reception if the quantity received is less than the original purchase order's quantity. The error message states: "Mismatch between move lines with the corresponding."

Steps to Reproduce:

Settings:

In General Settings (both companies):
Enable Sales from Purchase Orders.
Disable Sales Order Auto Validation.
In Purchase Settings: Enable Dropshipping.
In Inventory Settings: Enable Multi-Step Routes.
Setup:

In Company 1:
Configure a product with vendor information linked to Company 2.
In Company 2:
Configure the same product with vendor information linked to any other vendor (not a company contact).
Scenario:

In Company 1:
Create a sales order with a line using the Dropshipping route and the configured product.
Confirm the sales order.
Navigate to Purchase Orders, find the related record, and confirm it.
Switch to Company 2:
Locate the generated sales order and update the sales line to use the Dropshipping route.
Set the quantity to 2 and confirm the sales order.
Go to Purchase Orders, find the related record, and confirm it.
Navigate to Receipts for the purchase order, enter the lot/serial numbers, and attempt to validate the receipt.
Result: An error is raised, preventing the validation of the receipt due to a mismatch between the move lines and their corresponding records.

pedrobaeza commented on purchase_sale_stock_inter_company/models/stock_picking.py:
```diff
@@ -19,26 +19,61 @@ def action_done(self):
+            po_picking_pending = purchase.picking_ids.filtered(
+                lambda x: x.state not in ["done", "cancel"]
+            )
+            po_picking_pending.write({"intercompany_picking_id": pick.id})
```
Although it gets the same result, I consider dot notation cleaner:

```suggestion
            po_picking_pending.intercompany_picking_id = pick.id
```

and why writing this on each loop? It seems a bit weird this structure.

carlos-lopez-tecnativa replied:
Sorry, but I don’t understand. I only adapted the current code, but I don’t know the reasons. I think the action_done method can expect many records, and it’s better to iterate in a for loop.

pedrobaeza commented on purchase_sale_stock_inter_company/models/stock_picking.py:
```diff
@@ -19,26 +19,61 @@ def action_done(self):
-                            "corresponding  PO %s for assigning "
-                            "quantities and lots from %s for product %s"
+                po_move_lines = po_moves_pending.mapped("move_line_ids")
+                if move.product_id.tracking == "serial":
```
Why doing a special case depending on the product tracking instead of just mimicking the `stock.move.line`?

carlos-lopez-tecnativa replied:
> Why doing a special case depending on the product tracking instead of just mimicking the `stock.move.line`?

My error occurs only with serial products, so I isolated the error. In other cases, the check is fine, but for serial products, it’s not.

pedrobaeza replied:
I don't think so, as lots may have the same problem if you have more than one lot. As said, you should mimic the `stock.move.line` in one place and the other.

carlos-lopez-tecnativa replied:
Done, please update your review

pedrobaeza reviewed (commented):
AFAIK, this change is not changing existing `stock.move.line` to adapt things like the quantity, so it's not a real synchronization like it should be.

carlos-lopez-tecnativa committed:
[IMP] purchase_sale_stock_inter_company: Synchronize move lines from the origin company to the destination company

The synchronization is done only when the picking is in the "Done" state because, before that, the lots are not yet synchronized.

carlos-lopez-tecnativa commented:
> AFAIK, this change is not changing existing `stock.move.line` to adapt things like the quantity, so it's not a real synchronization like it should be.

@pedrobaeza I updated the code to synchronize the `stock.move.line`. Could you review it again? Thanks in advance.

pedrobaeza commented on purchase_sale_stock_inter_company/models/stock_move_line.py:
```diff
@@ -0,0 +1,95 @@
+                move_line.move_id.sale_line_id.sudo().auto_purchase_line_id
+            )
+            if move_line.state == "done" and purchase_line_origin:
+                po_moves = purchase_line_origin.move_ids.filtered(
```
Is it assured here that only one `stock.move` is selected?

carlos-lopez-tecnativa replied:
You're right, now the moves are iterated and filtered by product. Thanks for the observation!

pedrobaeza commented on purchase_sale_stock_inter_company/models/stock_move_line.py:
```diff
@@ -0,0 +1,95 @@
+                )
+                if not po_moves:
+                    continue
+                po_move_line_vals = po_moves._prepare_move_line_vals()
```
Why not simply copying this sml values and changing the `lot`? Anyway, I think there's an argument for passing the quantity directly to the method.

pedrobaeza replied:
And on write, you may need also to create some sml. Why not unifying the code?

carlos-lopez-tecnativa replied:
The argument `quantity` is used to populate the `product_uom_qty` field. In this case, I need to populate the `qty_done` field.
[Reference](https://github.com/odoo/odoo/blob/4cc086d330b73514fbc64a7fcf22d8a7a9f1b691/addons/stock/models/stock_move.py#L1105).

In the `write` method, the corresponding `stock.move.line` in the other company is always updated. There's no need to create a new record. Note that this code applies only to moves in the done state.

pedrobaeza replied:
And shouldn't you populate also that field?

carlos-lopez-tecnativa replied:
Done

pedrobaeza commented on purchase_sale_stock_inter_company/models/stock_move_line.py:
```diff
@@ -0,0 +1,95 @@
+    def create(self, vals_list):
+        new_move_lines = super().create(vals_list)
+        for move_line in new_move_lines:
+            purchase_line_origin = (
```
Is the direction OK from sales to purchase instead of the reverse?

carlos-lopez-tecnativa replied:
The direction is indeed from the sale order to the purchase order.

In this OCA module, the flow typically starts with a purchase order, which creates a new sale order. However, the delivery picking is always directed from the sale order to the purchase order.

In our case, for the customer, the flow begins with a sale order, continues to a purchase order (via Dropshipping), and then, from the purchase order, a new sale order is created according to the intercompany module. Hence, this direction is correct.

pedrobaeza replied:
But the use case is not important here. Having MTO doesn't matter. The question is that you should receipt the purchase, and that synchronize the sales, not the contrary.

carlos-lopez-tecnativa replied:
In this case, the user must fill in the lot in the sale order of the company where the delivery is made. If I sync from the purchase to the sale, how can I determine which product lot it is? Sorry, but I don’t understand your point.

pedrobaeza replied:
Yes, my fault, you are right. Delivery picking triggers the synchronization to receipt picking.

victoralmau reviewed (approved):
Code review OK

pedrobaeza commented:
/ocabot merge patch
